### ISSUE TITLE
Default number of jobs is unsafe (cabal-install)

### ISSUE THREAD

**Issue by hasufell** [haskell/cabal#5776]
Just ran `cabal new-install pandoc-include-code` without really thinking and it blew up my 16gb ram machine very quickly. Earlyoom was even unable to kill any of the processes, so I had to wait 15 minutes for the machine to become responsive again, potentially getting file corruption and break my recent work.

Number of processors is almost always the wrong choice for jobs with GHC. This is something that comes from the gcc world and even there it's sometimes wrong if the objects are too big.

The default should be **safe**, **not** *fast*. We shouldn't make random assumptions about the memory of the users machine.

**Comment by alanz** [comment]
The file `~/.cabal/config` has an entry for the number of CPUs to use, by default all of them

```
-- jobs: $ncpus
```

I change this to a fixed number, normally 80% or so of the available CPUs.

```
jobs: 6
```

**Comment by 23Skidoo** [comment]
`0.8 * $ncpus` might indeed be a better default.

**Comment by hasufell** [comment]
> `0.8 * $ncpus` might indeed be a better default.

That is still a guess. The default should be plain and simple: 1.

**Comment by phadej** [comment]
I don't see how setting `jobs: 1` would prevent a single GHC process from eating all the RAM anyway: less luckily, but still possible.

**Comment by hasufell** [comment]
> I don't see how setting `jobs: 1` would prevent a single GHC process from eating all the RAM anyway: less luckily, but still possible.

It is quite unlikely and has not happened a single time over half a year with projects being compiled regularly: pandoc, amazonka, gogol, docusign-base.

However, with the default, chances are 90% my laptop crashes or the OOM killer is triggered if I am lucky.

> Forcing `jobs: 1` by default feels like throwing out the baby with the bathwater to me; there must be a better heuristic for inferring a safe upper bound of parallelism (e.g. by factoring in the amount of available memory; possibly combined with passing RTS flags to `ghc`) to reduce the risk of an OOM than to force >90% of cabal users to tweak their configuration cause `jobs:1` is certainly not what the vast majority of cabal users wants... ;-)

You're not really "forcing" anything here. Defaults are not meant to be smart, but safe. If the user sees slow compile times, he may tweak the setting. Until anyone can come up with a safe heuristic (which is quite unlikely, because there are too many variables), I would say the default should be **1**.

**Comment by phadej** [comment]
Out of curiosity, what's your $ncpus (you have 16GB of RAM, don't you?)

**Comment by hasufell** [comment]
> Out of curiosity, what's your $ncpus (you have 16GB of RAM, don't you?)

8 cores
16GB of ram

**Comment by amigalemming** [comment]
I am annoyed for long by the ubiquitous assumption that $NCPUS is a good default choice for the number of threads, not only in Cabal. Since the advent of multitasking systems ressources must be shared between different processes. Since then it was not ok to eat up all memory by a single application and it was no longer ok to do busy waits, because there are other programs running on the machine and now for the same reasons it is not ok to occupy all cores without permission.
On my machine the number of cores is virtually 8 because of hyperthreading. But the hyperthreads practically don't do any speedup, they only eat up memory and make harddisk go crazy. Even 4 cores are not the optimum because of heavy harddisk load. i usually choose -j2 and --ghc-option=-j2.
A Raspberry Pi has 4 cores and 1 GB of RAM. With the default "jobs" configuration you turn quickly into thrashing.
Please stop eating up all my machine ressources without asking, please, please, please!

**Comment by gbaz** [comment]
I'd be keen on `0.5 * ncpus` as a good default, even on a desktop machine where a user is perhaps multitasking while compiling.

**Comment by gbaz** [comment]
As noted above, `jobs: 1` is not what the majority of users would want, so we should not cater to a small niche case above and beyond good defaults for the majority.

**Comment by hasufell** [comment]
> As noted above, `jobs: 1` is not what the majority of users would want, so we should not cater to a small niche case above and beyond good defaults for the majority.

How do you know what the majority of users want? I think you can safely say the majority of users do not want to lose control over their machine. IF you can come up with a safe default that takes into account users current load, available memory and so on, please go ahead. Otherwise I think it is unreasonable and dangerous to set jobs to arbitrary values. Both ncpu and 0.5 * ncpu are arbitrary.

**Comment by gbaz** [comment]
Yes. All values are arbitrary! But we have to pick one :-) After all, it is only a default, and it is easy to change.

**Comment by hasufell** [comment]
> But we have to pick one :-)

Yes and we should pick a safe one.

**Comment by hasufell** [comment]
> I think what you really want is a mechanism to control _memory_ usage

No no no. This is all really not a problem. This is solely about "I ran cabal new-install on a new machine and it crashed it". The result is just surprising, because the default is unsafe. After discovering it, everyone already knows how to deal with it. But it's horrible usability to crash a users machine in order to tell him to change his config.

**Comment by gbaz** [comment]
Right, and if cabal could detect when a machine did not have sufficient memory to run a parallel build, then that would avoid this!

**Comment by hasufell** [comment]
> Right, and if cabal could detect when a machine did not have sufficient memory to run a parallel build, then that would avoid this!

I think such proposals are out of scope of this issue and are not controversial. If someone can come up with such a solution, it can be done. But that should not block this issue from first making the default safe, until someone has a better solution.

**Comment by amigalemming** [comment]
> Right, and if cabal could detect when a machine did not have sufficient memory to run a parallel build, then that would avoid this!

I am still concerned with the attitude "We take as much as we can of a certain ressource without asking the user". Perhaps this attitude is based on the assumption that if the user starts the program then this program shall be the main task of the machine. Problem: There are many such programs. It is pretty common that I run Cabal in the background and want to use some cores for other purposes. It is also pretty common that Cabal makes my machine unresponsive by excessive harddisk accesses. Think also of a cluster machine where you have tons of cores but also many users. It is simply not ok to occupy many cores by default there.

And then there is the important argument: KISS. It is a pretty simple solution to set the default jobs to 1 and let the user increase it until he reaches the optimum for his machine. Any solution that tries to be more clever than the user will certainly be pretty complicated, requires APIs that are not available everywhere and rely on assumptions that still do not apply universally (e.g. "memory is the constraining factor for using many threads").

**Comment by tom-audm** [comment]
I've also never hit this problem at home or at work and would personally rather have people opt out of `$ncpus` when necessary - and have massively faster builds by default - than to restrict the number of CPUs (which, as noted, is really a fuzzy proxy for RAM usage anyway) to prevent the rare case of severe RAM contention.

I would, though, be in favor of @gbaz 's suggestion of "throttling" cabal (e.g. by not creating more GHC subprocesses) if system memory is low.
